### ISSUE TITLE
Allow capital letters in context urlPath

### ISSUE THREAD

**Issue by asmecher** [pkp/pkp-lib#6105]
See https://forum.pkp.sfu.ca/t/error-configuracion-de-la-ruta-de-la-revista-al-actualizar-a-ojs-3-2-1-1/61803 for a report/use case.

**Comment by asmecher** [comment]
@NateWr, see https://github.com/pkp/pkp-lib/commit/2e9aea7cfbc75c93015aaa38194fe791e2517f91 -- I'm assuming this omission was unintentional, but if not, please let me know!

**Comment by NateWr** [comment]
Thanks Alec. Does this create the potential for non-unique context paths, though? My understanding is that some servers don't differentiate between `/publicknowledge` and `/publicKnowledge` in a URL. With this change wouldn't it be possible to to have two different contexts at what the server considers the same URL?

It seems like most servers accept upper and lower-case letters in URLs, and users may expect that behaviour too. For example, you can access this issue at `https://github.com/PKP/PKP-LIB/issues/6097` and it redirects...

**Comment by asmecher** [comment]
@NateWr, as far as I know server paths are generally case-sensitive, though client implementations can choose to discard that if they want. From the issue, previous versions of OJS used to allow mixed case but since the move to the schema-based regexp for validation they no longer are. This puts existing users in a tough spot. I don't see any downside to allowing mixed cases again.

**Comment by asmecher** [comment]
(I suppose another approach would be to force lowercase on the form, but change the router to accept capital-letter variations. I'm open to the idea, but it would require some invasive changes.)

**Comment by NateWr** [comment]
I suspect this is one of those things where we give people room to hang themselves, but probably not worth upending things too much for now. Let's leave this as-is.

**Comment by NateWr** [comment]
What about allowing upper- or lower-case letters but throwing an error if two context paths match regardless of case? This should allow the old upper-case paths but prevent silly situations with duplicate paths.

**Comment by NateWr** [comment]
I think we could match path names with case-insensitivity here and maybe get the best of both worlds? https://github.com/pkp/pkp-lib/blob/782a53441a3a232e1bec5f8bd6f46f4bfe98e660/classes/context/ContextDAO.inc.php#L61

**Comment by asmecher** [comment]
> I think we could match path names with case-insensitivity here and maybe get the best of both worlds?

Good idea -- but I'm still smelling a tar pit. I googled around a bit and found [this](http://www.thesempost.com/using-caps-urls-not-bad-google-seo/) (maybe outdated?)

> So Google doesn’t consider letter case in itself as an SEO issue, but you should keep it consistent across individual URLs, because Google considers URLs to be case-sensitive.  This means, if you have the same page accessible and indexable (without canonicals) as example.com/PAGE.html and example.com/page.html, Google could crawl both those URLs and the one you want to be ranked may not be the one Google chooses.  You could also end up splitting ranking signals between the two pages.

**Comment by NateWr** [comment]
Ok good to know. Let's leave it as-is.

**Comment by asmecher** [comment]
Thanks, @NateWr!

**Comment by carzamora** [comment]
Hi, 

When I try to change the URL Path (automatically was written with some capital letters), I can't save because I have this error. Is the same problem described in this issue?

![image](https://user-images.githubusercontent.com/5422984/94099975-aa5f8d80-fe02-11ea-9e4d-166ca8f83879.png)

**Comment by NateWr** [comment]
Yes, you will need to update to 3.2.1 or later to be use capital letters in the URL Path.

**Comment by carzamora** [comment]
Thanks! But, my version is 3.2.1.1... I think this issue was opened after the release of this version; maybe I need to update from GitHub now, right?

**Comment by NateWr** [comment]
You're right. It looks like this went in just after 3.2.1.1 was released. We're planning to get a 3.2.1-2 out soon. Until then you'll need to patch it manually. It's a pretty small change: https://github.com/pkp/pkp-lib/commit/cc6cc9c48d0adfbdb3affa9f8ffc40ee36727795

**Comment by carzamora** [comment]
Thank you!

**Comment by spppaperview** [comment]
The changes to the JSON files are not enough to allow uppercase characters to be submitted in the URL Path. There is additional validation in the Form PHP files:

controllers/grid/articleGalleys/form/ArticleGalleyForm.inc.php
controllers/grid/issues/form/IssueGalleyForm.inc.php

We needed to change both for Article and Issue URL paths.

**Comment by NateWr** [comment]
@spppaperview you're right, this change only effects the `urlPath` of a context (ie - journal, press or preprint server). I'd be willing to open up galley `urlPaths` as well. Would you be willing to create a PR with the changes to those forms?

**Comment by spppaperview** [comment]
Hello @NateWr,
I'm afraid I have no dev experience -- I had to look up what a PR is. Hahaha! I'll see if I can figure out how to make one this weekend. But I would definitely like to become more involved in the future, we're benefiting so much from the OJS community.

In the mean time, I'll paste the code snippet here for the forum readers. I inserted the characters A-Z in the following Form Validator for urlPath in both IssueGalleyForm.inc.php and ArticleGalleyForm.inc.php:

`$this->addCheck(new FormValidatorRegExp($this, 'urlPath', 'optional', 'validator.alpha_dash', '/^[-_a-zA-Z0-9]*$/'));`

Cheers!

> @spppaperview you're right, this change only effects the `urlPath` of a context (ie - journal, press or preprint server). I'd be willing to open up galley `urlPaths` as well. Would you be willing to create a PR with the changes to those forms?
